rcl/test/rcl/test_events.cpp

```diff
@@ -0,0 +1,546 @@
+      liveliness_policy)) <<
+      "Initialized subscriber RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE when unsupported";
+    EXPECT_EQ(RMW_RET_ERROR,
+      setup_publisher(deadline, lifespan, lease_duration,
+      liveliness_policy)) <<
+      "Initialized publisher RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE when unsupported";
+
+    liveliness_policy = RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC;
+    EXPECT_EQ(RMW_RET_ERROR,
+      setup_subscriber(deadline, lifespan, lease_duration,
+      liveliness_policy)) <<
+      "Initialized subscriber RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC when unsupported";
+    EXPECT_EQ(RMW_RET_ERROR,
+      setup_publisher(deadline, lifespan, lease_duration,
+      liveliness_policy)) <<
+      "Initialized publisher RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC when unsupported";
+  }
+}
+
+TEST_F(CLASSNAME(TestEventFixture, RMW_IMPLEMENTATION), test_unsupported_unsupported_deadline) {
+  if (is_unsupported) {
+    rmw_time_t deadline {1, 0};
+    rmw_time_t lifespan {0, 0};
+    rmw_time_t lease_duration {0, 0};
+    rmw_qos_liveliness_policy_t liveliness_policy = RMW_QOS_POLICY_LIVELINESS_AUTOMATIC;
+    EXPECT_EQ(RMW_RET_ERROR,
+      setup_subscriber(deadline, lifespan, lease_duration,
+      liveliness_policy)) << "Initialized subscriber deadline when unsupported";
+    EXPECT_EQ(RMW_RET_ERROR,
+      setup_publisher(deadline, lifespan, lease_duration,
+      liveliness_policy)) << "Initialized publisher deadline when unsupported";
+
+    deadline = {0, 1};
+    EXPECT_EQ(RMW_RET_ERROR,
+      setup_subscriber(deadline, lifespan, lease_duration,
+      liveliness_policy)) << "Initialized subscriber deadline when unsupported";
+    EXPECT_EQ(RMW_RET_ERROR,
+      setup_publisher(deadline, lifespan, lease_duration,
+      liveliness_policy)) << "Initialized publisher deadline when unsupported";
+  }
+}
+
+/*
+ * Basic test of publisher and subscriber liveliness events, with publisher killed
+ */
+TEST_F(CLASSNAME(TestEventFixture, RMW_IMPLEMENTATION), test_pubsub_liveliness_kill_pub)
+{
+  if (is_unsupported) {
+    return;
+  }
+  setup_publisher_and_subscriber(RCL_PUBLISHER_LIVELINESS_LOST, RCL_SUBSCRIPTION_LIVELINESS_CHANGED);
+  rcl_ret_t ret;
+
+  // publish message to topic
+  const char * test_string = "testing";
+  {
+    test_msgs__msg__Primitives msg;
+    test_msgs__msg__Primitives__init(&msg);
+    ASSERT_TRUE(rosidl_generator_c__String__assign(&msg.string_value, test_string));
+    ret = rcl_publish(&publisher, &msg);
+    test_msgs__msg__Primitives__fini(&msg);
+    ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  }
+
+  // kill the publisher
+  ret = rcl_event_fini(&publisher_event);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  ret = rcl_publisher_fini(&publisher, this->node_ptr);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+  // wait for lease duration to expire
+  std::this_thread::sleep_for(LIVELINESS_LEASE_DURATION_IN_S + milliseconds(500));
```

Comment:
I would like to see all sleeps gone if possible.

Instead, I think it would be better if `wait_for_msgs_and_events` would check the status during this time (even though we expect it not to change), and even compare it to a timestamp taken before the `fini` to make sure it doesn't expire early. Expiring shortly after the lease duration is ok, but before should be a bug (in my opinion), similar to the contract given by `sleep_for` (i.e. it may sleep >= period, but not less than).

Reply:
Testing for the "happens too early" case might be tricky, so you don't have to do that, but I'd still rather just have it sitting in `rcl_wait()` than sleeping here.